src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -0,0 +1,431 @@
+                    {
+                        break;
+                    }
+
+                    bytesRead += b;
+                    bytesLeft -= b;
+                }
+
+                AndroidTzDataHeader header = LoadHeader(buffer);
+                ReadIndex(fs, header.indexOffset, header.dataOffset);
+            }
+
+            private unsafe AndroidTzDataHeader LoadHeader(Span<byte> buffer)
+            {
+                // tzdata files are expected to start with the form of "tzdata2012f\0" depending on the year of the tzdata used which is 2012 in this example
+                // since we're not differntiating on year, check for tzdata and the ending \0
+                Span<byte> signature = buffer.Slice(0, 12);
+                Span<byte> tzDataSignature = stackalloc byte[8];
+                tzDataSignature.Clear();
+
+                signature.Slice(0, 6).CopyTo(tzDataSignature);
+                ulong tzdata = (ulong)TZif_ToInt64(tzDataSignature);
+
+                char[] cbits = Encoding.Default.GetChars(tzDataSignature.ToArray());
+
+                string sigBits = "";
+                for (int i = 0; i < signature.Length; i++)
+                {
+                    sigBits += signature[i].ToString() + " ";
+                }
+
+                if (tzdata != 0x747A646174610000 || signature[11] != 0)
+                {
+                    // 0x747A646174640000 = {0x74, 0x7A, 0x64, 0x61, 0x74, 0x64, 0x00, 0x00} = "tzdata[NUL][NUL]"
+                    var b = new StringBuilder(signature.Length);
+                    for (int i = 0; i < signature.Length; ++i)
+                    {
+                        b.Append(' ').Append(HexConverter.ToCharLower(signature[i]));
+                    }
+
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_BadTZHeader, TimeZoneFileName, b.ToString()));
+                }
+
+                AndroidTzDataHeader header;
+
+                header.indexOffset = TZif_ToInt32(buffer.Slice(12, 4));
+                header.dataOffset = TZif_ToInt32(buffer.Slice(16, 4));
+                header.finalOffset = TZif_ToInt32(buffer.Slice(20, 4));
+
+                return header;
+            }
+
+            private unsafe void ReadIndex(Stream fs, int indexOffset, int dataOffset)
+            {
+                int indexSize = dataOffset - indexOffset;
+                int entrySize = sizeof(AndroidTzDataEntry);
+                int entryCount = indexSize / entrySize;
+
+                _byteOffsets = new int[entryCount];
+                _ids = new string[entryCount];
+                _lengths = new int[entryCount];
+
+                for (int i = 0; i < entryCount; ++i)
+                {
+                    AndroidTzDataEntry entry = LoadEntryAt(fs, indexOffset + (entrySize*i));
+                    var p = (sbyte*)entry.id;
+
+                    _byteOffsets![i] = entry.byteOffset + dataOffset;
+                    _ids![i] = new string(p);
```

Comment:
The docs say this uses the Encoding.Default. Is that always guaranteed to be correct here? Or should we instead be using `Encoding.ASCII.GetString(bytes)`? (or maybe UTF8? whichever the Android docs say this string is encoded in)

Reply:
It originally was converted to use Encoding.ASCII in the mono/mono implementation. There was feedback to use `string(p)` https://github.com/dotnet/runtime/pull/54845#discussion_r668562095

It's not clear to me if we need to use ASCII encoding or why that was added in mono/mono, and i havent been able to find much documentation on `tzdata` format itself besides trying to understand the bionic files https://github.com/aosp-mirror/platform_bionic/blob/master/libc/tzcode/bionic.cpp

Reply:
From what I've seen, it's ascii.  I think we should be explicit like mono/mono is.

Reply:
ASCII encoding is subset of UTF8. You can think about ASCII encoding as UTF8 with more error handling.

If you assume that the file format is valid, both ASCII and UTF8 are going to work well here.

The downside of using ASCII encoding is that it is rarely used, and so it may have negative size impact on small trimmer apps.

Reply:
I would assume ASCII as well, given they are using [`strcmp in the bionic code.`](https://github.com/aosp-mirror/platform_bionic/blob/2e4be55966d0eb6d71260aa65b445141821644de/libc/tzcode/bionic.cpp#L178)

Reply:
strcmp works just fine for UTF8 too.